common/src/androidTest/java/com/microsoft/identity/common/SharedPreferencesBrokerApplicationMetadataCacheTest.java

```diff
@@ -86,6 +100,106 @@ public void testInsert() {
         );
     }
 
+    @Test
+    public void testMultipleInsertOrUpdate() throws Exception {
+        final ExecutorService executor = Executors.newFixedThreadPool(10);
+        final CountDownLatch latch =  new CountDownLatch(1);
+        final CountDownLatch executionLatch =  new CountDownLatch(10);
+        final Set<BrokerApplicationMetadata> set = Collections.synchronizedSet(new HashSet<BrokerApplicationMetadata>());
+        List<Future<Boolean>> list = new ArrayList<>();
+        for (int i = 0; i <10; i++){
+            list.add(executor.submit(new Callable<Boolean>() {
+                @Override
+                public Boolean call() throws Exception {
+                    BrokerApplicationMetadata metadata = generatePredictiveMetadata();
+                    set.add(metadata);
+                    return insertOrUpdateHelper(metadata, latch, executionLatch);
+                }
+            }));
+        }
+        executionLatch.await();
+        latch.countDown();
+
+        for(Future<Boolean> f : list){
+            Assert.assertTrue(f.get());
+        }
+
+        final int size = mMetadataCache.getAll().size();
+
+        assertEquals("Expected 1 record, but got too many",
+                1,
+                size
+        );
+
+        Assert.assertTrue(set.contains(mMetadataCache.getAll().get(0)));
+    }
+
+    public boolean insertOrUpdateHelper
+            (BrokerApplicationMetadata metadata, CountDownLatch latch, CountDownLatch executionLatch)
+            throws InterruptedException {
+        executionLatch.countDown();
+        latch.await();
+        return mMetadataCache.insertOrUpdate(metadata);
+    }
+
+    @Test
+    public void testReadsWithInsertOrUpdate() throws Exception {
```

Comment:
I think the is the seconds test I discussed with you.  The thing is, you kind of want to test multiple possible interleaving of this mess, and there's actually not a really good way to control it.  The best we're going to be able to do is get a "probably covered it".  Meaning that you want to set this up to strobe the cache structure until you tell it to stop - probably a set number of iterations or total ops per reader/writer pool, like 100000 or so.

So you don't need to have all the latching, necessarily, though controlling the start doesn't hurt you.  You will want to execute your read in a loop, potentially with a micro/milli second sleep in there to make sure you don't starve the writers while you continually insert/update the data store.